### PR TITLE
Collision check

### PR DESCRIPTION
**Pull request by dalinel**
This PR adds a collision check inside the controller ```solve()``` function
The ```CollisionCheck``` class has been implemented. It needs a collision yaml per robot
The yaml defines spheres per members
Spheres are described by their diameter and relative position to a frame

The collision check takes a pinocchio model in a posture ```q``` and checks if members are colliding
If a model collision is detected inside the controller  ```send_cmd_``` and ```q_``` is not updated anymore
The robot will stop moving

The total computation time of such collision check is of around 50 microseconds on talos

After a collision the qp will keep following the references and solving the problem
If you need to send commands again after a collision you can use the ```set_send_cmd``` 
The  ```qp_step_back``` function can be used to put the qp back to a previous state

### CODE REVIEW

File: include/inria_wbc/controllers/controller.hpp [hunk 1]
```diff
@@ -160,6 +161,14 @@ namespace inria_wbc {
             const std::string& urdf() const { return urdf_; }
             const std::string& floating_base_joint_name() const { return floating_base_joint_name_; }
 
+            //check if pinocchio model is colliding
+            inria_wbc::utils::CollisionCheck collision_check() { return collision_check_; }
+            bool is_model_colliding() { return is_model_colliding_; }
+            void set_send_cmd(const bool& send_cmd) { send_cmd_ = send_cmd; };
```
Comment (ivanbergonzani) [inline include/inria_wbc/controllers/controller.hpp:167]: no need for const reference on bool

File: include/inria_wbc/robot_dart/utils.hpp [hunk 2]
```diff
@@ -40,6 +47,65 @@ namespace inria_wbc {
             return new_commands;
         }
 
+        //draw utils for model collisions
+        inline std::vector<std::shared_ptr<::robot_dart::Robot>> create_spherical_members(
+            const std::map<std::string, std::vector<std::pair<Eigen::Vector3d, float>>>& spherical_members,
```
Comment (jbmouret) [inline include/inria_wbc/robot_dart/utils.hpp:52]: I think we should move this to a .cpp

File: include/inria_wbc/utils/collision_check.hpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    class CollisionCheck {
+
+    public:
+        CollisionCheck(){};
+
+        //Load the collision spheres yaml.
+        //A sphere is represented by a 4d float array [x_relative, y_relative, z_relative, diameter].
+        //The position of the sphere is relative to the position of the frame name in the yaml file.
+        void load_collision_file(const std::string& collision_file_path, bool verbose = false)
```
Comment (jbmouret) [inline include/inria_wbc/utils/collision_check.hpp:19]: We should move to a .cpp too.

File: src/controllers/controller.cpp [hunk 4]
```diff
@@ -274,6 +306,10 @@ namespace inria_wbc {
                 error += " (t=" + std::to_string(t_) + ")";
                 throw IWBC_EXCEPTION(error);
             }
+            if (check_model_collisions_)
+                is_model_colliding_ = collision_check_.is_colliding(robot_->model(), tsid_->data());
+            if (is_model_colliding_ && !q_.isZero(1e-3))
+                send_cmd_ = false;
```
Comment (jbmouret) [inline src/controllers/controller.cpp:312]: In verbose mode we could add a print that there is a collision?

**Comment by jbmouret** [comment]
I would like to have a safe mode where we stop the robot much before the actual collision.

Proposal : 
- add a parameter to increase the margin (like, 20%) so that we are in "safe mode"
- look at collisions at t=0, put them in a list, and ignore them (so that we can can deal with the shoulder intersecting with the torso in an automatic way).

The alternative is to have two set of spheres, one "tight" and one "safe", that we hand-optimize. What do you think?

File: etc/talos/talos_pos_tracker.yaml [hunk 5]
```diff
@@ -29,6 +29,8 @@ CONTROLLER:
                      "gripper_right_inner_single_joint",
                      "gripper_right_fingertip_3_joint",
                      "gripper_right_motor_single_joint"]
+  check_model_collisions: true
+  collision_path: /home/pal/inria_wbc/etc/talos/collisions/talos_collisions.yaml
```
Comment (jbmouret) [inline etc/talos/talos_pos_tracker.yaml:33]: We should not use an absolute path here (like, it does not work on my computer).

Comment (jbmouret) [reply etc/talos/talos_pos_tracker.yaml:33]: (we should use the base_path parameter)

File: src/controllers/controller.cpp (hunk 4 above)
Comment (dalinel) [reply src/controllers/controller.cpp:312]: There  is already this print inside ```CollisionCheck::is_colliding```

**Comment by dalinel** [comment]
Ok I have made the corrections
For the margin I have made a new collision yaml adding 5cm diameter margin on the arms only

**Pull request merged**